linting/src/no_main.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use ast::{
+    AttrStyle,
+    Crate,
+};
+use clippy_utils::diagnostics::span_lint_and_help;
+use if_chain::if_chain;
+use rustc_ast as ast;
+use rustc_lint::{
+    EarlyContext,
+    EarlyLintPass,
+    LintContext,
+};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{
+    declare_lint,
+    declare_lint_pass,
+};
+use rustc_span::sym;
+
+declare_lint! {
+    /// ## What it does
+    /// Checks if a contract is annotated with the `no_main` outer attribute.
+    ///
+    /// ## Why is this necessary?
+    /// Contracts must be annotated with `no_main` outer attribute when compiled to WebAssembly for
```

Comment:
Not for Wasm, in fact, the `no_main` is required to compile to `riscv` targets. So you could just say e.g. to allow for compilation to all supported targets.